src/QsCompiler/Core/Transformations/ExpressionTransformation.fs

```diff
@@ -17,51 +17,36 @@ open Microsoft.Quantum.QsCompiler.Transformations.Core.Utils
 
 type private ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>
 
+type ExpressionKindTransformationBase(expressionTransformation: _ -> ExpressionTransformationBase, options) =
+    static let createExpressionTransformation kinds options =
+        let types = TypeTransformationBase options
+        let expressions = ExpressionTransformationBase((fun () -> kinds), (fun () -> types), options)
+        fun () -> expressions
 
-type ExpressionKindTransformationBase internal (options: TransformationOptions, _internal_) =
+    let node = if options.Rebuild then Fold else Walk
 
-    let missingTransformation name _ =
-        new InvalidOperationException(sprintf "No %s transformation has been specified." name) |> raise
+    member _.Types = expressionTransformation().Types
 
-    let Node = if options.Rebuild then Fold else Walk
+    member _.Expressions = expressionTransformation ()
 
-    member val internal ExpressionTransformationHandle = missingTransformation "expression" with get, set
-
-    member this.Expressions: ExpressionTransformationBase = this.ExpressionTransformationHandle()
-    member this.Types: TypeTransformationBase = this.ExpressionTransformationHandle().Types
-    member this.Common: CommonTransformationNodes = this.ExpressionTransformationHandle().Types.Common
-
-    new(expressionTransformation: unit -> ExpressionTransformationBase, options: TransformationOptions) as this =
-        new ExpressionKindTransformationBase(options, "_internal_")
-        then this.ExpressionTransformationHandle <- expressionTransformation
+    member internal this.Common = expressionTransformation().Types.Common
```

Comment:
External transformations should access the methods through the `SyntaxTreeTransformation` class. `ICommonTransformation` is an implementation detail to break the mutual recursion between the transformation types while they are being defined.

Reply:
If someone were to just create an `ExpressionTransformation`, how would they get access to `OnTypeRange` for example?

Reply:
If `this` is of type `ExpressionTransformation`, then `this.Transformation.OnTypeRange` should work.

Reply:
Oh all the non-Base transformation classes have a reference to a SyntaxTreeTransformation object with those handles. That addresses how I can _call_ the OnTypeRange from an Expression Transformation. How would I overwrite the OnTypeRange though?

Reply:
With the current implementation, I can define a class that inherits from `CommonTransformationNode` and overwrite the `OnTypeRange` behavior. Then I could define my nuanced `ExpressionTransformation` class and assign it to my `CommonTransformationNode` class by assigning `this.Common = MyCommonTransformationNode;` in the constructor.

Reply:
I think there is a misunderstanding. This PR doesn't change the way you override the common transformation methods. In main as well as in this PR, you can only override them in SyntaxTreeTransformation.

> With the current implementation, I can define a class that inherits from `CommonTransformationNode` and overwrite the `OnTypeRange` behavior.

CommonTransformationNodes cannot be inherited from because its constructor is internal. If you try to inherit from it, you get this error: "[CS1729] 'CommonTransformationNodes' does not contain a constructor that takes 0 arguments"

> Then I could define my nuanced `ExpressionTransformation` class and assign it to my `CommonTransformationNode` class by assigning `this.Common = MyCommonTransformationNode;` in the constructor.

You cannot assign to Common because it is a read-only property. If you try to assign a different instance of CommonTransformationNode, you get this error: "[CS0200] Property or indexer 'ExpressionTransformationBase.Common' cannot be assigned to -- it is read only"